### ISSUE TITLE
UdpClient hangs when Receive is waiting and gets a Close() in macos

### ISSUE THREAD

**Issue by jaimefinat** [dotnet/runtime#64551]
Hi
we are experiencing an issue in macos (Monterey) when closing a UdpClient socket from the server side.
Also related to previous detected issue [#21392](https://github.com/dotnet/runtime/issues/24513)

1. We create the UdpClient: 
`mUdpClient = new UdpClient(listenPort);`
2. The UdpClient calls Receive method:
`byte[] bytes = mUdpClient.Receive(ref groupEP);`
3. While waiting to receive something from any client, the Close() method is called in another thread:
mUdpClient.Close();

Attached to this, I'm sending a minimum version of a Server-Client application to reproduce the issue. 

Expected behaviour:
In Windows it is rising an exception which allos us to stop the server:
```
System.Net.Sockets.SocketException (10004): A blocking operation was interrupted by a call to WSACancelBlockingCall.
   at System.Net.Sockets.Socket.ReceiveFrom(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint& remoteEP)
   at System.Net.Sockets.UdpClient.Receive(IPEndPoint& remoteEP)
   at Server.StartListener() in C:\experiments\client-server\server\Server.cs:line 21
```

In macos it hangs in Receive until it receives a message from the client, so it process the request and then triggers another exception:
```
Unhandled exception. System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
   at System.Net.Sockets.Socket.ThrowObjectDisposedException()
   at System.Net.Sockets.Socket.ReceiveFrom(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint& remoteEP)
   at System.Net.Sockets.UdpClient.Receive(IPEndPoint& remoteEP)
   at Server.StartListener() in /Users/daniel.hompanera/wkspaces/UdpApp/Server/Server.cs:line 21
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
[1]    29462 abort      dotnet Server.dll
```

To reproduce the issue (different behaviour in macos/windows) download and build the solution from this post, start the server and type: _efss_ to EndFromServerSide.

The code has been modified from the basic version [here](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/using-udp-services)

Thank you very much for your kind support!
Jaime

[UdpApp.zip](https://github.com/dotnet/runtime/files/7973222/UdpApp.zip)

**Comment by jaimefinat** [comment]
We are using a timeout, as the async alternative would imply some more refactoring and redesign. I hope it is aso OK.

Thank you for your quick response!
Best,
Jaime

**Comment by wfurt** [comment]
Understand. However, I'm not sure if there is any good fox for this case.
You could use something like
```c#
using CancellationTokenSource cts = new CancellationTokenSource();
cts.CancelAfter(timeout);
var result = socket.ReceiveAsync(memory, flags, cts.Token).GetAwaiter().GetResult();
```
While this is generally not recommended it should avoid the hang. 
The exception would be different.

Actually digging deeper this seems like dup of https://github.com/dotnet/runtime/issues/47342

https://github.com/dotnet/runtime/blob/a2235e775b55efab0e195d2850f1b4bb12c8a4f1/src/libraries/System.Net.Sockets/tests/FunctionalTests/ReceiveFrom.cs#L174-L181

**Comment by tmds** [comment]
It's a known limitation that we're unable to unblock sync receives and TCP connects:

https://github.com/dotnet/runtime/blob/daf77fc9a5ee0417d691a19322cc0bc74d8562df/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L189-L190

You need to use an async API if you want to be able to cancel them.

**Comment by wfurt** [comment]
I was thinking about it more @tmds and I'm wondering if we can do something like:
```
 tmp = handle;
 handle = -1;
 close(tmp);
```

e.g. set handle to invalid to prevent future operations or mixup when the handle is closed but actually use `close()` to finalize all pending operations.  

The FIN behavior does not matter for UDP and may be acceptable for TCP in corner cases. Still feels better than blocking forever.

**Comment by tmds** [comment]
On macOS, when there is a pending UDP `receive`, `close` does not cause it to return.
It's not something we're causing, it is how the kernel behaves.

**Comment by wfurt** [comment]
I don't think that is true @tmds. Consider following example
```c#
using System;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using System.Runtime.InteropServices;

namespace udpListener
{
    class Program
    {
        [DllImport("libc", EntryPoint = "close", SetLastError = true)]
        static extern internal int MyClose(IntPtr handle);

        static void Receive(Socket s)
        {
            try
            {
                byte[] b = new byte[10024];
                int received = s.Receive(b);
                Console.WriteLine("Received {0} bytes", received);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }
        static void Main(string[] args)
        {
            int listenPort = 9000;
            var mUdpClient = new UdpClient(listenPort);
            Console.WriteLine("Got listening socket on {0}", mUdpClient.Client.Handle);
            var t = Task.Run(() => Receive(mUdpClient.Client));
            Thread.Sleep(1000);
            Console.WriteLine("Cancelling");

            if (args.Length > 0)
            {
                MyClose(mUdpClient.Client.Handle);
            }
            else
            {
                mUdpClient.Client.Close();
            }

            t.GetAwaiter().GetResult();
            Console.WriteLine("All done");
        }
    }
}
```
It hangs when executed without parameters but
```
Shining:udpListener furt$ ~/dotnet-6/dotnet run a
Hello World!
Got listening socket on 33
Cancelling
System.Net.Sockets.SocketException (89): Operation canceled
   at System.Net.Sockets.Socket.Receive(Byte[] buffer)
   at udpListener.Program.Receive(Socket s) in /Users/furt/projects/udpListener/Program.cs:line 20
All done
```

I think it is OK to close the handle while used in p/invoke but we need to avoid case when it is closed but old value could be used. This is why I suggested to set the handle to invalid to prevent further use before closing the handle.

**Comment by tmds** [comment]
> I don't think that is true

Yes, you're right. Calling `close` causes the `recv` to unblock.

> This is why I suggested to set the handle to invalid to prevent further use before closing the handle.

This has a race. When the handle gets closed between adding the reference to the `SafeHandle` and passing the fd to the syscall, the fd passed to the syscall may no longer refer to the right socket.

**Comment by wfurt** [comment]
> This has a race. When the handle gets closed between adding the reference to the `SafeHandle` and passing the fd to the syscall, the fd passed to the syscall may no longer refer to the right socket.

can we fix that somehow? I'm not sure if setting the handler to invalid is sufficient but it can be start.

**Comment by tmds** [comment]
> can we fix that somehow? I'm not sure if setting the handler to invalid is sufficient but it can be start.

No. This is the race:

```
Thread A: takes ref on safe handle
Thread A: int fd = safehandle.DangerousGetHandle();
Thread B: close(fd)
Thread B: int s = socket(); // the raw handle 's' equals the 'fd'
Thread A: recvmsg(fd); // recvmsg was supposed to be called on safehandle, but it is called on 's' instead.
```

**Comment by tipa** [comment]
I am using a Socket (in order to get an NTP timestamp) on macOS and I'm hitting two problems:

#### 1) Cancelling either `ConnectAsync`, `SendAsync` or `ReceiveAsync` doesn't seem to be working reliably
In my code (see code example below) I am using a cancellation token that should abort all socket operations after 1.5 seconds (and throw a `TaskCanceledException`). Therefore, the code shouldn't go into the `else`-branch that is only entered when my `Task.Delay` task, which is configured to wait 2.5 sec, is finishing earlier. But my logging shows that the code does enter that branch - even though `linked_tcs.IsCancellationRequested` is set to true. How is this possible? Is the `IsCancellationRequested` property not checked at least every second?

#### 2) Users are reporting that the app is hanging and unresponsive
Even though I am using async methods, users are reporting that the app is hanging/unresponsive and they have to force-quit it. Are there any blocking operatings that are executed in the main thread, even when using the async `Connect`, `Send` or `Receive` methods of the `Socket` class?

```
using var linked_tcs = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
linked_tcs.CancelAfter(1500); // only wait 1.5 sec for NTP timestamp
var delayTask = Task.Delay(2500, cancellationToken); // only wait 2.5 sec for Task.WhenAny to return
var queryTask = QueryAsync(linked_tcs.Token);
var winner = await Task.WhenAny(queryTask, delayTask).ConfigureAwait(false);
if (winner == queryTask)
{
    // all good, we got the response
}
else
{
    // we should never get here because linked_tcs is cancelled after 1.5 sec and Task.Delay() waits for 2.5 sec
    // linked_tcs.IsCancellationRequested is *true* here
    // cancellationToken.IsCancellationRequested is *false* here
}

public static async Task<Span<byte> buffer> QueryAsync(CancellationToken token = default)
{
    using var socket = new Socket(SocketType.Dgram, ProtocolType.Udp);
    await socket.ConnectAsync("pool.ntp.org", 123, token).ConfigureAwait(false);
    var rcvbuff = new byte[160];
    await socket.SendAsync(GetRequestBytes(), token).ConfigureAwait(false);
    int length = await socket.ReceiveAsync(rcvbuff, token).ConfigureAwait(false);
    return rcvbuff.AsSpan(0, length);
}
```

This problem only happens on macOS (but not on iOS, Android etc).
I'm using the latest .NET9 version, problem occurs on macOS 15.2.
If this is the wrong thread to post this in, I can also open a new issue
